src/objects.c

```diff
@@ -425,13 +425,14 @@ void PrepareCopy(Obj obj, Obj copy)
     // between original and copy (e.g. for objects, they point to the type;
     // if making an immutable copy of a mutable object, the types will
     // differ).
+    // Store in a DATOBJ, as CONST_ADDR_OBJ(obj)[0] could be anything,
+    // including an arbitary C integer.
     // Likewise, the TNUM of the copy and the original can and will differ;
     // e.g. for a weak pointer list, the copy can be a plist.
-    Obj tmp = NEW_PLIST(T_PLIST, 3);
-    SET_LEN_PLIST(tmp, 3);
-    SET_ELM_PLIST(tmp, 1, CONST_ADDR_OBJ(obj)[0]);
```

Comment:
For which kind of object is this not a valid object reference?

Reply:
It's hard to tell exactly what object, as I'm just seeing the objects later. I think it's C_STRINGs

Reply:
OK, indeed, for `T_STRING` the first word is the length of the string as a raw integer. I just went through all calls to `PrepareCopy` to check for each type using it what is going on:
* blist: first slot contains length as an intobj
* posobj, comobj, datobj: first slot contains ref to type object
* plist: first slot contains length as an intobj
* prec: first slot is always 0
* range: first slot contains length as an intobj
* string: first slot contains length as a **raw integer** (!!!)
* weak pointer object: first slot contains length as a **raw integer** (!!!)

So strings and wpobjs are the oddballs. I think for wpobj, it is an absolute nobrainer to switch to storing the length as an INTOBJ. This leaves strings. They also used to store the length as an intobj, but at some point this was changed (IIRC by Frank) to allow larger strings. This is only relevant on 32bit systems, though. Hence, at least on 64bit systems, we could change this to an intobj, too.

I'd thus suggest this fix: Leave this code here as it is. Change wpobj to use an intobj length. Change strings also to store the integer as an object, but with a special case for 32bit systems, where instead of `INTOBJ_INT` and `INT_INTOBJ`, we use the more general `ObjInt_UInt` resp. `UInt_ObjInt`.

Reply:
I don't think Frank changed them to be raw integers, I think he added the fact we stored the length at all, instead of assuming strings are always the same size as the bag they are contained in.

Reply:
I don't like the idea of using ObjInt_Uint, as that means that setting the length of a string could cause a GC, so we'd have to check everywhere in both the kernel, and packages, to make sure no-one is keeping a C pointer (which is a common thing to do when playing with strings) in such cases.

If we just switched to using INTOBJ_INT (and checked the maximum size of strings) that would limit us to 256MB strings on a 32-bit OS, which doesn't sound like an unreasonable requirement, but is possible to hit.

Reply:
Here is an alternate suggestion: just add another argument to `PrepareCopy`, a boolean indicating whether slot 0 contains raw data or an object reference. Store that boolean in the plist, too; and then, depending on its value, store slots 0 in the plist either verbatim, as we do now (if the boolean indicates it is an object ref), or else using `ObjInt_UInt` (if the boolean indicates it is raw data).

Reply:
Thanks for the suggestions.

I'm actually going to go and do some investigations. I've been poking both strings and blists, and it turns out neither of them work properly if you actually get larger than a value that fits in an INTOBJ, for example if you can make a blist and string of size >= 2^28, then `l[2^28]` doesn't work, and I can easily get GAP to crash by asking various questions about these 2^28 length blists and strings.

Reply:
To be concrete, I think your original idea is best - - make the lengths intobjs, as larger values are currently borken and that is the easier and more stable fix, in my mind.

Reply:
To be concrete, I think your original idea is best - - make the lengths intobjs and add error checking for lengths, as larger values are currently borken anyway.